File: libraries/oakestra_messaging_go/topics_test.go
package messaging

import (
	"testing"

	"gotest.tools/v3/assert"
)

func TestMatches(t *testing.T) {
	cases := []struct {
		name    string
		pattern string
		topic   string
		want    bool
	}{
		{"exact match", "nodes/n1/control/deploy", "nodes/n1/control/deploy", true},
		{"plus wildcard matches one segment", "nodes/+/information", "nodes/n1/information", true},
		{"plus does not swallow an extra segment", "nodes/+/information", "nodes/n1/extra/information", false},
		{"plus does not swallow a longer trailing segment", "nodes/+/net/subnet", "nodes/n1/net/subnetwork/result", false},
		{"foreign prefix does not match", "nodes/n1/control/deploy", "xnodes/n1/control/deploy", false},
		{"trailing extra segment does not match without hash", "nodes/n1/control/deploy", "nodes/n1/control/deploy/extra", false},
		{"unrelated prefix does not match", "nodes/n1/control/deploy", "other/n1/control/deploy", false},
		{"hash matches the exact prefix with zero remaining segments", "nodes/n1/#", "nodes/n1", true},
		{"hash matches multiple remaining segments", "nodes/n1/#", "nodes/n1/a/b", true},
		{"hash matches a single remaining segment", "nodes/n1/#", "nodes/n1/a", true},
		{"hash requires the prefix to match", "nodes/n1/#", "nodes/n2/a", false},
		{"hash not at the end never matches", "nodes/#/control", "nodes/n1/control", false},
		{"hash not at the end never matches even the literal topic", "nodes/#/control", "nodes/#/control", false},
	}

	for _, tc := range cases {
		t.Run(tc.name, func(t *testing.T) {
			assert.Equal(t, Matches(tc.pattern, tc.topic), tc.want)
		})
	}
}

File: libraries/oakestra_messaging/tests/topics_test.py
import pytest
from oakestra_messaging import matches

CASES = [
    # exact match
    ("nodes/n1/information", "nodes/n1/information", True),
    ("nodes/n1/information", "nodes/n1/job", False),
    # "+" matches exactly one segment
    ("nodes/+/information", "nodes/n1/information", True),
    ("nodes/+/information", "nodes/n1/extra/information", False),
    ("nodes/+/net/subnet", "nodes/n1/net/subnet", True),
    ("nodes/+/net/subnet", "nodes/n1/net/subnetwork/result", False),
    ("nodes/+/job", "nodes/n1/job", True),
    ("nodes/+/job", "job", False),
    # "#" matches the rest, including zero remaining segments
    ("nodes/n1/#", "nodes/n1", True),
    ("nodes/n1/#", "nodes/n1/net/subnet", True),
    ("nodes/n1/#", "nodes/n1/net/tablequery/result", True),
    ("#", "anything/at/all", True),
    ("#", "", True),
    # multiple wildcards; "+" is exactly one segment, so a deeper suffix
    # such as "tablequery/request" (two segments) does not match
    ("nodes/+/net/+", "nodes/n1/net/subnet", True),
    ("nodes/+/net/+", "nodes/n1/net/tablequery/request", False),
    ("nodes/+/net/+", "nodes/n1/other/subnet", False),
    # foreign prefixes never match
    ("nodes/+/information", "jobs/j1/updates_available", False),
    ("jobs/+/updates_available", "nodes/n1/information", False),
    ("nodes/+/information", "other/n1/information", False),
    # pattern longer or shorter than topic (no "#")
    ("nodes/+/information", "nodes/n1", False),
    ("nodes/n1", "nodes/n1/information", False),
]


@pytest.mark.parametrize("pattern, topic, expected", CASES)
def test_matches(pattern, topic, expected):
    assert matches(pattern, topic) is expected
